### DISCUSSION TITLE
Hashing enums

### DISCUSSION THREAD

**Discussion by mnijhuis-tos** [wjakob/nanobind#105] [Q&A] [answered]
While converting pybind11 bindings for an enumeration type, I noticed the following:
- Simply replacing py:: by nb:: was enough. The existing value() and export_values() calls still worked. Great!
- When using a set with enumeration values, I got an error that the type is not hashable. Unlike pybind11, nanobind does not add a `__hash__` function to the type.

When adding a `__hash__` function myself, I noticed the following:
- Since nb::enum_ is derived from nb::class_, I can use the `def(__hash__, ...)`. However, the return value of `def()` is an `nb::class_`, and not an `nb::enum_`, so I had to put this `def()` after all the `value()` calls.
- Adding the `def("__hash__", ...)` indeed removes the error that the type is not hashable.
- Instead of the not-hashable error, I'm now getting a "RuntimeError: nanobind::detail::nb_func_vectorcall_simple(): the 'self' argument of a method call should be a nanobind class." My function looks as follows:
`.def("__hash__", [](const Enum& e) { return static_cast<int>(e); })`
`Enum` is the same type as in the `nb::enum_<Enum>()` call.

Can you tell me if you plan to support hashing enums, like pybind11 does, and/or tell me if and how I can add my own `__hash__` function to an `nb::enum_` binding?

**Comment by wjakob** [comment, marked as answer]
Hi @mnijhuis-tos -- that's a good catch. Could you create a PR that implements the hashing function in the libnanobind component? (specifically, that would be in the file `src/nb_enum.cpp`). Instead of using `.def("__hash__", ...)` would do this at a lower level by providing the `Py_tp_hash` type slot in this file. Hopefully it's clear enough if you look at the other code there.

**Reply by mnijhuis-tos** [reply]
I've written an nb_enum_hash function, based on nb_enum_int. See https://github.com/wjakob/nanobind/pull/106 .
